### PR TITLE
WIP Add uint8 support for interpolate on channels_last (1, 3, H,W) CPU images, mode=bilinear, antialias=True

### PR DESCRIPTION
**Pull request by NicolasHug**
This is heavily adapted / ported from PILLOW-SIMD.

This PR adds support for `uint8` images in the following case:
- mode=bilinear, `antialias=True` -- support for other modes is possible
- shape=(1, 3, H, W)  -- can extend to batch_size > 1. Not sure about channels yet.
- layout = channels_last  --- we could easily support contiguous as well, since we have to copy (pack / unpack) the input anyway
- device=CPU

This may sound restrictive, but it's not. This is exactly the setting in which torchvision' `Resize()` is used for training jobs.

This is still WIP with lots of TODOs, but it seems to be working decently. On the inputs I tried and comparing with torchvision's `Resize()` (which first converts uint8 to floats, runs `interpolate()`, and converts back to uint8), I'm getting ~3-5X speedup. It seems correct so far as the absolute difference in the outputs is never `> 1`, and only ~15% of the pixel values differ (by exactly 1).

This addresses https://github.com/pytorch/vision/issues/2289

@vfdev-5 @mingfeima @fmassa I'd love your initial thoughts on this!

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Is it vectorizing the reads/writes? If so, may also be nice to support 4-channel inputs (RGBA?) or any divisible by 4/8 inputs - probably vectorization is even simpler in this case

**Comment by mingfeima** [comment]
Is this a migration of PIL-simd's kernel? Shall we parallel it at the same time, shouldn't be too much additional job ~

**Comment by NicolasHug** [comment]
Thanks for the feedback @vadimkantorov. Yes, the vectorized part is  roughly `output[i] = sum_j (wj * input[j])`. The pre-computation of the weights and the index mapping isn't vectorized. Regarding the next feature to support, we could add these to our backlog for sure. I'll have to gauge interest internally to decide what to prioritize though.

@mingfeima yes this is a direct port from PIL-SIMD. Indeed I think we should be able to parallelize over the batch-size. Did you have something else in mind?

**Comment by vadimkantorov** [comment]
Maybe I wasn't clear, I meant is the read-ins and write-outs themselves vectorized? Are all three R,G,B channels read and written in one go? For uint8 3-channels, maybe they just fit in one uint32, but for 3-channels float32, SSE-registers may be used. The memory access would probably be not very aligned, as the memory is read by triplets (and not quadruplets), but maybe in modern processors it's not hurting much.

**Comment by NicolasHug** [comment]
Hm, I'm not sure what you mean by read-ins and write-outs honestly. If you mean the [packing/unpacking](https://github.com/pytorch/pytorch/pull/87863/files#diff-79c0925a5e86d740f34d20ab3c660c503f5220c2316d384a0a11f3fcc2c04962R45-R72) of the input/output, that part isn't vectorized. The vectorized part is the writing to the unpacked output from the unpacked input. 

> Are all three R,G,B channels read and written in one go? 

Yes.

**Comment by vadimkantorov** [comment]
I guess packing/unpacking part wasn't very clear. Is it first copying the memory in unpacked format and then processing it in vectorized way? I guess my question is whether this first unpacking is needed and can be fused with processing (but this would incur some useless computation for non-existing 4-th channel).

**Comment by NicolasHug** [comment]
> Is it first copying the memory in unpacked format and then processing it in vectorized way

Yes, the input tensor arrives as `R G B R G B R G B...` where  each letter is a uint8. For the vectorized code to run, we have to first unpack all RGB triplets into 32bits (the last 8 bits are set to 255). So this becomes `R G B P R G B P ...` where `P` is the 255 padding.

This does involve extra copies, and there might be opportunities to improve this in the future. But for now, even with the extra copies this is still way faster than `torchvision.Resize()`, and better than no uint8 support at all; so it's definitely worth it.

**Comment by vadimkantorov** [comment]
Then yes, maybe in the futures copies can be avoided and can be fused directly with reading (probably worth adding a todo or explanation about this in the code...)

**Comment by vadimkantorov** [comment]
hope soon pillow is not needed for simple pipelines :)

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 1]
```diff
@@ -1376,8 +1377,22 @@ void upsample_bilinear2d_aa_kernel_impl(
     c10::optional<double> scales_h,
     c10::optional<double> scales_w) {
 
-  separable_upsample_generic_Nd_kernel_impl<2, scale_t, HelperInterpLinear>(
-      output, input, align_corners, {scales_h, scales_w});
+  if (input.dtype() == at::kByte) {
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:1380]: you might want to use something like `_use_vectorized_kernel_cond` here (a few functions in the file seem to be using that condition to decide whether or not to vectorize)

File: aten/src/ATen/native/cpu/my_interp.h [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+    // /* none of the previous steps are performed, copying */
+    // if (!imOut) {
+    //     imOut = ImagingCopy(imIn);
+    // }
+
+    return unpacked_output_p;
+}
+
+void beepidiboop(const Tensor& input, const Tensor& output) {
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/my_interp.h:340]: You should add a template for scalar type

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 3]
```diff
@@ -1376,8 +1377,22 @@ void upsample_bilinear2d_aa_kernel_impl(
...
+    // - Use dispatcher instead
+    // - Put the proper guards in place: this only works for (1, 3, H, W) images that are channel_last
+    // - Re-organize files
+    // - Use vec.h instead of hardcoded AVX2 intrinsics
+    // - As much as possible, re-use existing code e.g. for coeff computation
+    // - Use pytorch types instead of UINT32, INT16 etc.
+    // - There's a segfault when input_shape == output_shape
+    // - This could be extended to other filters, not just bilinear
+    // - License?
+    beepidiboop(input, output);
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:1391]: call this with the appropriate `AT_DISPATCH_{}` macro once the function is templated

File: aten/src/ATen/native/cpu/resample_simd_vertical.c [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+#include "blah.h"  // TODO: remove
+
+// TODO: Use stuff in vec.h instead
+namespace {
+
+void ImagingResampleVerticalConvolution8u(
+    UINT32 *lineOut, UINT32 * imIn,
+    int xmin, int xmax, INT16 *k, int coefs_precision, int xin)
+{
+#ifdef CPU_CAPABILITY_AVX2
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/resample_simd_vertical.c:11]: add a comment sharing the link from where this was borrowed

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp (hunk 1 above)
Comment (anjali411) [reply aten/src/ATen/native/cpu/UpSampleKernel.cpp:1380]: synced offline -- currently the focus is to only support this for uint8. In the future, we can extend this to other data types.
